core/management/commands/setup_index.py

```diff
@@ -0,0 +1,130 @@
+import glob
+import logging
+import os
+import requests
+from urllib import parse
+
+from django.core.management.base import BaseCommand
+from django.conf import settings
+from core.management.commands import configure_logging
+from solr import SolrConnection
+
+configure_logging('setup_index_logging.config', 'setup_index.log')
+
+_logger = logging.getLogger(__name__)
+fixture_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../fixtures'))
+schema_url = settings.SOLR_API + '/cores/openoni/schema'
+
+# Copy fields are defined here because we have to manually check for dupes; for
+# some reason Solr doesn't do this for us, and will in fact allow dozens of the
+# same copy-field definition.  The structure should be obvious, and is the
+# exact format Solr's API takes.
+copy_fields = [
+    {'source': 'place_of_publication', 'dest': 'place_of_publication_facet'},
+    {'source': 'subject', 'dest': 'subject_facet'},
+    {'source': 'title', 'dest': 'title_facet'},
+    {'source': 'ocr_*', 'dest': '_text_'}
+]
+
+class Command(BaseCommand):
+    help = 'Set up all Solr index configuration necessary for Open ONI'
+
+    def handle(self, **options):
+        _logger.info('Loading Solr schema fixtures')
+
+        # Schema fixtures must be idempotent!  Nothing here should cause
+        # problems if run multiple times.
+        schema_fixtures = sorted(glob.glob(os.path.join(fixture_dir, 'solr-schema', '*.json')))
+        for fixture in schema_fixtures:
+            if not self.load_schema_fixture(fixture):
+                _logger.error('Exiting...')
+                return
+
+        _logger.info('Adding copy fields')
+        existing_copy_fields = self.get_existing_copy_fields()
+        for definition in copy_fields:
+            if not self.defined_copy_field(existing_copy_fields, definition):
+                if not self.create_copy_field(definition):
+                    _logger.error('Exiting...')
+                    return
+
+    def get_existing_copy_fields(self):
+        with requests.get(schema_url) as r:
+            if r.status_code != 200:
+                _logger.error('Unable to read schema from Solr: ' + r.text)
+                _logger.error('Exiting...')
+                return
+
+            schema = r.json()['schema']
+            if 'copyFields' in schema:
+                return schema['copyFields']
+
+    def defined_copy_field(self, existing_copy_fields, definition):
```

Comment:
I don't think I have a particular order per se - I tried to group functions somewhat logically, but it's not like there's a hard and fast rule I went by.  As an example, the bit that loads all the fixtures is followed immediately by all the various functions that help it out (`load_schema_fixture`, `valid_response`, `valid_add_field_type_response`, and `valid_add_field_response).  Pretty much everything except `handle` could probably be private, but even then I'd argue for keeping them grouped by what part of the process they handle: the fixture loading versus the more manual copy-field defining bits.

As far as I know, there are no guidelines that suggest a specific ordering of functions.  I'd prefer to not try and establish one.  If a file is so huge that it's difficult to find functions, that's could be a sign that we need to look into breaking things apart, but there are certainly valid cases where we need a billion functions, and it may make a lot more sense for things to be logically grouped, such as having something like `transform_data` next to `process_data` rather than alphabetized.

If you feel the file is getting too complex or has too many unrelated functions, rather than alphabetize I'd think we should consider extracting some of the code into more servicey bits, like a SolrSchemaFixtureLoader and a SolrCopyFieldCreator or something.  I personally think that's overkill, as this file is pretty small still.

Reply:
There doesn't seem to be anything in the Python style guide about this, so we have nothing to reference there or in our own dev guidelines right now. So I say we leave this for another day.

When we get around to general code cleanup someday, I would be interested in defining some organization preferences we can apply across the code base and have as a reference for code review beyond initial cleanup.